dashboard-widgets.el

```diff
@@ -633,23 +652,119 @@ WIDGET-PARAMS are passed to the \"widget-create\" function."
       (insert (propertize footer 'face 'dashboard-footer))
       (insert "\n"))))
 
+;;
+;; Truncate
+;;
+(defun dashboard-f-filename (path)
+  "Return file name from PATH."
+  (file-name-nondirectory path))
+
+(defun dashboard-f-base (path)
+  "Return directory name from PATH."
+  (file-name-nondirectory (directory-file-name (file-name-directory path))))
+
+(defun dashboard-shorten-path-beginning (path)
+  "Shorten PATH from beginning if exceeding maximum length."
+  (let* ((len-path (length path)) (len-rep (length dashboard-path-shorten-string))
+         (len-total (- dashboard-path-max-length len-rep))
+         front)
+    (if (<= len-path dashboard-path-max-length) path
+      (setq front (substring path (- len-path len-total) len-path))
+      (concat dashboard-path-shorten-string front))))
+
+(defun dashboard-shorten-path-middle (path)
+  "Shorten PATH from middle if exceeding maximum length."
+  (let* ((len-path (length path)) (len-rep (length dashboard-path-shorten-string))
+         (len-total (- dashboard-path-max-length len-rep))
+         (center (/ len-total 2))
+         (end-back center)
+         (start-front (- len-path center))
+         back front)
+    (if (<= len-path dashboard-path-max-length) path
+      (setq back (substring path 0 end-back)
+            front (substring path start-front len-path))
+      (concat back dashboard-path-shorten-string front))))
+
+(defun dashboard-shorten-path-end (path)
+  "Shorten PATH from end if exceeding maximum length."
+  (let* ((len-path (length path)) (len-rep (length dashboard-path-shorten-string))
+         (len-total (- dashboard-path-max-length len-rep))
+         back)
+    (if (<= len-path dashboard-path-max-length) path
+      (setq back (substring path 0 len-total))
+      (concat back dashboard-path-shorten-string))))
+
+(defun dashboard-shorten-path (path)
+  "Shorten the PATH."
+  (setq path (abbreviate-file-name path))
+  (cl-case dashboard-path-style
+    (truncate-beginning (dashboard-shorten-path-beginning path))
+    (truncate-middle (dashboard-shorten-path-middle path))
+    (truncate-end (dashboard-shorten-path-end path))
+    (t path)))
+
+(defun dashboard-shorten-paths (paths alist)
+  "Shorten all path from PATHS and store it to ALIST."
+  (let (lst-display abbrev (index 0))
+    (setf (symbol-value alist) nil)  ; reset
+    (dolist (item paths)
+      (setq abbrev (dashboard-shorten-path item)
+            ;; Add salt here, and use for extraction.
+            ;; See function `dashboard-extract-key-path-alist'.
+            abbrev (format "%s|%s" index abbrev))
```

Comment:
Here is quite complicated, by adding the index and "|" to prevent having the same id (abbrev).